examples/peps/utils.jl

```diff
@@ -0,0 +1,222 @@
+)
+  o⃗ψ = ψ
+  for oᵢ in o⃗
+    o⃗ψ = apply(oᵢ, o⃗ψ; cutoff, maxdim, normalize)
+  end
+  return o⃗ψ
+end
+
+function flattened_inner_network(ϕ::ITensorNetwork, ψ::ITensorNetwork)
+  tn = inner(ϕ, ψ)
+  for v in vertices(ψ)
+    tn = ITensors.contract(tn, (2, v...) => (1, v...))
+  end
+  return tn
+end
+
+function contract_inner(ϕ::ITensorNetwork, ψ::ITensorNetwork; sequence=nothing)
+  tn = inner(ϕ, ψ)
+  # TODO: convert to an IndsNetwork and compute the contraction sequence
+  for v in vertices(ψ)
+    tn = ITensors.contract(tn, (2, v...) => (1, v...))
+  end
+  if isnothing(sequence)
+    sequence = optimal_contraction_sequence(tn)
+  end
+  return ITensors.contract(tn; sequence)[]
+end
+
+norm2(ψ::ITensorNetwork; sequence) = contract_inner(ψ, ψ; sequence)
+
+function ITensors.expect(op::String, ψ::ITensorNetwork; cutoff, maxdim)
+  res = Dictionary(vertices(ψ), Vector{Float64}(undef, nv(ψ)))
+  sequence = optimal_contraction_sequence(flattened_inner_network(ψ, ψ))
+  normψ² = norm2(ψ; sequence)
+  for v in vertices(ψ)
+    O = ITensor(Op(op, v), s)
+    Oψ = apply(O, ψ; cutoff, maxdim)
+    res[v] = contract_inner(ψ, Oψ; sequence) / normψ²
+  end
+  return res
+end
+
+function ITensors.expect(ℋ::OpSum, ψ::ITensorNetwork; cutoff, maxdim)
+  s = siteinds(ψ)
+  h⃗ = Vector{ITensor}(ℋ, s)
+  sequence = optimal_contraction_sequence(flattened_inner_network(ψ, ψ))
+  h⃗ψ = [apply(hᵢ, ψ; cutoff, maxdim) for hᵢ in h⃗]
+  ψhᵢψ = [contract_inner(ψ, hᵢψ; sequence) for hᵢψ in h⃗ψ]
+  ψh⃗ψ = sum(ψhᵢψ)
+  ψψ = norm2(ψ; sequence)
+  return ψh⃗ψ / ψψ
+end
+
+function randomITensorNetwork(s; link_space)
+  ψ = ITensorNetwork(s; link_space)
+  for v in vertices(ψ)
+    ψᵥ = copy(ψ[v])
+    randn!(ψᵥ)
+    ψᵥ ./= norm(ψᵥ)
+    ψ[v] = ψᵥ
+  end
+  return ψ
+end
+
+function ITensors.MPO(opsum::OpSum, s::IndsNetwork)
+  s_linear = [only(s[v]) for v in 1:nv(s)]
+  return MPO(opsum, s_linear)
+end
+
+function ITensors.randomMPS(s::IndsNetwork, args...; kwargs...)
+  s_linear = [only(s[v]) for v in 1:nv(s)]
+  return randomMPS(s_linear, args...; kwargs...)
+end
+
+function ITensors.MPS(s::IndsNetwork, args...; kwargs...)
+  s_linear = [only(s[v]) for v in 1:nv(s)]
+  return MPS(s_linear, args...; kwargs...)
+end
+
+maybe_only(x) = x
+maybe_only(x::Tuple{T}) where {T} = only(x)
+
+function ising(g::AbstractGraph; h)
+  ℋ = OpSum()
+  for e in edges(g)
+    ℋ -= "Z", maybe_only(src(e)), "Z", maybe_only(dst(e))
+  end
+  for v in vertices(g)
+    ℋ += h, "X", maybe_only(v)
+  end
+  return ℋ
+end
+
+function tebd(ℋ::OpSum, ψ::ITensorNetwork; β, Δβ, maxdim, cutoff)
+  𝒰 = exp(-Δβ * ℋ; alg=Trotter{2}())
+  # Imaginary time evolution terms
+  u⃗ = Vector{ITensor}(𝒰, s)
+  nsteps = Int(β ÷ Δβ)
+  for step in 1:nsteps
+    if step % 10 == 0
+      @show step, (step - 1) * Δβ, β
+    end
+    ψ = ITensorNetworks.insert_links(ψ)
+    ψ = apply(u⃗, ψ; cutoff, maxdim, normalize=true)
+  end
+  return ψ
+end
```

Comment:
You should be able to delete this file now (this code was moved into the package).

Reply:
Done